File: src/main/java/com/github/highright1234/glacier/protocol/handler/ReceivingEventHandler.java
package com.github.highright1234.glacier.protocol.handler;

import com.github.highright1234.glacier.EventManager;
import com.github.highright1234.glacier.protocol.MinecraftPacket;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelInboundHandlerAdapter;
import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
public class ReceivingEventHandler extends ChannelInboundHandlerAdapter {

    private final EventManager eventManager;

    @Override
    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
        if (msg instanceof MinecraftPacket) {
            eventManager.callEvent(msg);
        }
    }
}

File: src/main/kotlin/com/github/highright1234/glacier/protocol/handler/ReceivingEventHandler.kt
package com.github.highright1234.glacier.protocol.handler

import com.github.highright1234.glacier.protocol.MinecraftPacket
import kotlin.Throws
import java.lang.Exception
import io.netty.channel.ChannelInboundHandlerAdapter
import io.netty.channel.ChannelHandlerContext
import com.github.highright1234.glacier.event.EventManager
import com.github.highright1234.glacier.protocol.PipelineUtil
import com.github.highright1234.glacier.event.event.PacketReceivingEvent

data class ReceivingEventHandler(val eventManager : EventManager) : ChannelInboundHandlerAdapter() {

    @Throws(Exception::class)
    override fun channelRead(ctx: ChannelHandlerContext, msg: Any) {
        if (msg is MinecraftPacket) {
            if (PipelineUtil.getClientConnection(ctx.channel())?.let {
                    PacketReceivingEvent(it, msg)
                }?.let {
                    eventManager.callEvent( it ).isCancelled
                }!!
            ) {
                ctx.fireChannelRead(msg)
            }
        }
    }
}